README.md
# Squeak

[![build_badge]][build_link] [![crates.io_badge]][crates.io_link] [![docs_badge]][docs_link]

[build_badge]: https://img.shields.io/github/actions/workflow/status/agersant/squeak/CI.yml?branch=master
[build_link]: https://github.com/agersant/squeak/actions/workflows/CI.yml?query=branch%3A+branch%3Amaster++
[crates.io_badge]: https://img.shields.io/badge/crates.io-squeak-green
[crates.io_link]: https://crates.io/crates/squeak
[docs_badge]: https://img.shields.io/badge/docs.rs-squeak-blue
[docs_link]: https://docs.rs/squeak/latest/squeak/

Squeak is a zero-dependency Rust library to facilitate event-driven programming.

# Examples

```rust
use squeak::{Delegate, Response};

let on_damage_received = Delegate::new();
on_damage_received.subscribe(|amount| {
    println!("Received {amount} damage");
    Response::StaySubscribed
});

on_damage_received.broadcast(16); // Prints "Received 16 damage"
on_damage_received.broadcast(14); // Prints "Received 14 damage"
on_damage_received.broadcast(28); // Prints "Received 28 damage"
```

```rust
use squeak::{Observable, Response};

let mut health = Observable::new(100);
health.subscribe(|updated_health| {
    println!("Health is now {updated_health}");
    Response::StaySubscribed
});

health.mutate(|h| *h -= 10); // Prints "Health is now 90"
health.mutate(|h| *h -= 5);  // Prints "Health is now 85"
health.mutate(|h| *h += 25); // Prints "Health is now 110"
```


CHANGELOG.md
# Changelog

## Version 0.2.0

- Squeak is now a `no_std` crate.

## Version 0.1.0

- Initial release.


tests/test_delegate.rs
use parking_lot::{Mutex, ReentrantMutex};
use std::{cell::RefCell, ops::Deref, sync::Arc};

use squeak::{Delegate, Response};

#[test]
fn delegate_executes_callbacks() {
    let mut call_count = 0;
    {
        let d = Delegate::new();
        d.subscribe(|_| {
            call_count += 1;
            Response::StaySubscribed
        });
        d.notify();
        d.notify();
        d.notify();
    }
    assert_eq!(call_count, 3);
}

#[test]
fn can_subscribe_during_callback() {
    let d = Arc::new(ReentrantMutex::new(Delegate::new()));
    let outer_count = Arc::new(Mutex::new(RefCell::new(0)));
    let inner_count = Arc::new(Mutex::new(RefCell::new(0)));
    {
        let d_clone = d.clone();
        let outer_count_clone = outer_count.clone();
        let inner_count_clone = inner_count.clone();
        d.lock().subscribe(move |_| {
            *outer_count_clone.lock().borrow_mut() += 1;
            let inner_count_clone_clone = inner_count_clone.clone();
            d_clone.lock().subscribe(move |_| {
                *inner_count_clone_clone.lock().borrow_mut() += 1;
                Response::CancelSubscription
            });
            Response::CancelSubscription
        });
        d.lock().notify();
        d.lock().notify();
    }
    assert_eq!(*outer_count.lock().borrow(), 1);
    assert_eq!(*inner_count.lock().borrow(), 1);
}

#[test]
fn delegate_does_not_execute_unsubscribed_callbacks() {
    let mut call_count = 0;
    {
        let d = Delegate::new();
        let subscription = d.subscribe(|_| {
            call_count += 1;
            Response::StaySubscribed
        });
        d.unsubscribe(subscription);
        d.notify();
    }
    assert_eq!(call_count, 0);
}

#[test]
fn cannot_unsubscribe_using_subscription_from_a_different_delegate() {
    let mut call_count = 0;
    {
        let d1 = Delegate::<()>::new();
        let d2 = Delegate::<()>::new();
        let _s1 = d1.subscribe(|_| {
            call_count += 1;
            Response::StaySubscribed
        });
        let s2 = d2.subscribe(|_| Response::StaySubscribed);
        d1.unsubscribe(s2);
        d1.notify();
    }
    assert_eq!(call_count, 1);
}

#[test]
fn unsubscribing_within_callback_is_noop() {
    let d = Arc::new(ReentrantMutex::new(Delegate::new()));
    let call_count = Arc::new(Mutex::new(RefCell::new(0)));
    let subscription = Arc::new(Mutex::new(RefCell::new(None)));

    let d_clone = d.clone();
    let call_count_clone = call_count.clone();
    let subscription_clone = subscription.clone();

    subscription
        .lock()
        .replace(Some(d.lock().subscribe(move |_| {
            let old_count = *call_count_clone.lock().borrow();
            *call_count_clone.lock().borrow_mut() = old_count + 1;
            if let Some(subscription) = subscription_clone.lock().deref().borrow_mut().take() {
                d_clone.lock().unsubscribe(subscription);
            }
            Response::StaySubscribed
        })));

    d.lock().notify();
    d.lock().notify();
    assert_eq!(*call_count.lock().borrow(), 2);
}

#[test]
fn can_unsubscribe_using_response_value() {
    let mut call_count = 0;
    {
        let d = Delegate::new();
        d.subscribe(|_| {
            call_count += 1;
            Response::CancelSubscription
        });
        d.notify();
        d.notify();
    }
    assert_eq!(call_count, 1);
}


tests/test_observable.rs
use squeak::{Observable, Response};

#[test]
fn observable_broadcasts_new_values() {
    let mut seen_value = 0;
    {
        let mut o = Observable::new(0);
        o.subscribe(|new_value| {
            seen_value = *new_value;
            Response::StaySubscribed
        });
        o.mutate(|value| *value = 42);
    }
    assert_eq!(seen_value, 42);
}

#[test]
fn observable_no_longer_notifies_after_unsubscribe() {
    let mut call_count = 0;
    {
        let mut o = Observable::new(0);
        let s = o.subscribe(|_| {
            call_count += 1;
            Response::StaySubscribed
        });
        o.mutate(|value| *value = 42);
        o.unsubscribe(s);
        o.mutate(|value| *value = 43);
    }
    assert_eq!(call_count, 1);
}


src/delegate.rs
use alloc::vec::Vec;
use alloc::{borrow::Borrow, boxed::Box, collections::BTreeMap, fmt::Debug};

use core::cell::RefCell;
use core::sync::atomic::{AtomicU64, Ordering};

type BoxedCallback<'a, T> = Box<dyn FnMut(&T) -> Response + 'a + Send>;
type SubscriptionId = u64;

static NEXT_SUBSCRIPTION_ID: AtomicU64 = AtomicU64::new(0);

/// Maintains a list of callbacks that can be explicitely triggered
/// by calling [`Delegate::broadcast`].
#[derive(Default)]
pub struct Delegate<'d, T> {
    pub(crate) subscriptions: RefCell<BTreeMap<SubscriptionId, BoxedCallback<'d, T>>>,
}

/// Represents a subscription created via [`Delegate::subscribe`] or [`Observable::subscribe`](crate::Observable::subscribe).
///
/// It can be passed to [`Delegate::unsubscribe`] or [`Observable::unsubscribe`](crate::Observable::unsubscribe) to cancel the subscription.
#[derive(Eq, Hash, PartialEq)]
pub struct Subscription {
    id: SubscriptionId,
}

/// Returned by [`Delegate`] and [`Observable`](crate::Observable) subscription callbacks.
/// Depending on the value returned, the subscription will stay active or be cancelled.
pub enum Response {
    StaySubscribed,
    CancelSubscription,
}

impl<'d, T> Delegate<'d, T> {
    pub fn new() -> Self {
        Self {
            subscriptions: RefCell::new(BTreeMap::new()),
        }
    }

    /// Registers a new callback that will be called when this delegate broadcasts
    /// a new value.
    ///
    /// ```rust
    /// use squeak::{Delegate, Response};
    ///
    /// let on_damage_received = Delegate::new();
    /// on_damage_received.subscribe(|amount| {
    ///     println!("Received {amount} damage");
    ///     Response::StaySubscribed
    /// });
    /// on_damage_received.broadcast(5); // Prints "Received 5 damage"
    /// ```
    ///
    /// The output of the callback function determines whether it will be called
    /// again when [`broadcast`] is called in the future.
    ///
    pub fn subscribe<C: FnMut(&T) -> Response + 'd + Send>(&self, callback: C) -> Subscription {
        let id = NEXT_SUBSCRIPTION_ID.fetch_add(1, Ordering::SeqCst);
        let subscription = Subscription { id };
        self.subscriptions
            .borrow_mut()
            .insert(subscription.id, Box::new(callback));
        subscription
    }

    /// Removes a callback that was previously registered.
    ///
    /// ```rust
    /// use squeak::{Delegate, Response};
    ///
    /// let on_damage_received = Delegate::new();
    /// let subscription = on_damage_received.subscribe(|amount| {
    ///     println!("Received {amount} damage");
    ///     Response::StaySubscribed
    /// });
    /// on_damage_received.broadcast(5); // Prints "Received 5 damage"
    /// on_damage_received.unsubscribe(subscription);
    /// on_damage_received.broadcast(10); // Does not print anything
    /// ```
    /// - Attempting to unsubscribe using a [`Subscription`] that was created by a different [`Delegate`] has no effect.
    /// - Attempting to unsubscribe a [`Subscription`] multiple times has no effect.
    /// - Attempting to unsubscribe from within callback function has no effect.
    pub fn unsubscribe(&self, subscription: Subscription) {
        self.subscriptions.borrow_mut().remove(&subscription.id);
    }

    /// Executes all registered callbacks, providing `value` as their argument.
    ///
    /// ```rust
    /// use squeak::{Delegate, Response};
    ///
    /// let on_renamed = Delegate::new();
    /// on_renamed.subscribe(|new_name: &String| {
    ///     println!("New name is {new_name}");
    ///     Response::StaySubscribed
    /// });
    /// on_renamed.broadcast(String::from("Lisa"));
    /// on_renamed.broadcast(&String::from("Trevor"));
    /// on_renamed.broadcast(&mut String::from("Jill"));
    /// ```
    pub fn broadcast<U: Borrow<T>>(&self, value: U) {
        let subscriptions_to_notify = self
            .subscriptions
            .borrow()
            .keys()
            .copied()
            .collect::<Vec<_>>();
        for subscription in subscriptions_to_notify {
            let (_, mut callback) = self
                .subscriptions
                .borrow_mut()
                .remove_entry(&subscription)
                .unwrap();
            match callback(value.borrow()) {
                Response::CancelSubscription => (),
                Response::StaySubscribed => {
                    self.subscriptions
                        .borrow_mut()
                        .insert(subscription, callback);
                }
            };
        }
    }
}

impl Delegate<'_, ()> {
    /// This convenience function broadcasts the unit type on delegates with no payload.
    ///
    /// ```rust
    /// use squeak::{Delegate, Response};
    ///
    /// let on_respawn = Delegate::new();
    /// on_respawn.subscribe(|_| {
    ///    println!("Respawned");
    ///    Response::StaySubscribed
    /// });
    /// on_respawn.notify();
    /// ```
    pub fn notify(&self) {
        self.broadcast(());
    }
}

impl<T> Debug for Delegate<'_, T>
where
    T: Debug,
{
    fn fmt(&self, f: &mut alloc::fmt::Formatter<'_>) -> alloc::fmt::Result {
        f.debug_struct("Delegate")
            .field(
                "subscriptions",
                &format_args!("{} active subscriptions", self.subscriptions.borrow().len()),
            )
            .finish()
    }
}


src/observable.rs
use alloc::fmt::Debug;
use core::ops::Deref;

use crate::{Delegate, Response, Subscription};

/// Wrapper type which owns a value and executes callbacks every time a call is made to mutate the value.
///
/// ``` rust
/// use squeak::{Observable, Response};
///
/// let mut health = Observable::new(100);
/// health.subscribe(|updated_health| {
///     println!("Health is now {updated_health}");
///     Response::StaySubscribed
/// });
///
/// health.mutate(|h| *h -= 10); // Prints "Health is now 90"
/// health.mutate(|h| *h -= 5);  // Prints "Health is now 85"
/// health.mutate(|h| *h += 25); // Prints "Health is now 110"
/// ```
///
/// Observables implement [`std::ops::Deref`], which means the inner value can be accessed
/// via `*my_observable`.
#[derive(Debug)]
pub struct Observable<'o, T> {
    value: T,
    delegate: Delegate<'o, T>,
}

impl<'o, T> Observable<'o, T> {
    /// Creates a new observable with an initial value
    ///
    /// ```rust
    /// use squeak::Observable;
    /// let name = Observable::new(String::from("DefaultName"));
    /// ```
    pub fn new(value: T) -> Self {
        Self {
            value,
            delegate: Delegate {
                subscriptions: Default::default(),
            },
        }
    }

    /// Registers a new callback that will be called when the value contained in this observable is mutated.
    ///
    /// ```rust
    /// use squeak::{Observable, Response};
    ///
    /// let mut health = Observable::new(100);
    /// health.subscribe(|updated_health| {
    ///     println!("Health is now {updated_health}");
    ///     Response::StaySubscribed
    /// });
    /// ```
    ///
    /// The output of the callback function determines whether it will be called
    /// again when [`broadcast`] is called in the future.
    pub fn subscribe<C: FnMut(&T) -> Response + 'o + Send>(&self, callback: C) -> Subscription {
        self.delegate.subscribe(callback)
    }

    /// Removes a callback that was previously registered.
    ///
    /// ```rust
    /// use squeak::{Observable, Response};
    ///
    /// let mut health = Observable::new(100);
    /// let subscription = health.subscribe(|updated_health| {
    ///     println!("Health is now {updated_health}");
    ///     Response::StaySubscribed
    /// });
    ///
    /// health.unsubscribe(subscription);
    /// ```
    pub fn unsubscribe(&self, subscription: Subscription) {
        self.delegate.unsubscribe(subscription);
    }

    /// Returns a reference to a delegate that will execute subscription functions
    /// when the observable is mutated. This is useful when writing a struct that has
    /// an observable member, but users of the struct should only have access to its
    /// value by subscribing.
    ///
    /// ```rust
    /// use squeak::{Delegate, Observable};
    ///
    /// struct MyStruct<'a> {
    ///   observe_only: Observable<'a, u32>,
    /// }
    ///
    /// impl<'a> MyStruct<'a> {
    ///   pub fn delegate(&self) -> &Delegate<'a, u32> {
    ///     self.observe_only.delegate()
    ///   }
    /// }
    ///
    /// ```
    ///
    pub fn delegate(&self) -> &Delegate<'o, T> {
        &self.delegate
    }

    /// Execute a function which may mutate the value contained in this observable.
    /// Subscription callbacks will be executed regardless of what happens inside
    /// the `mutation` function.
    ///
    /// ```rust
    /// use squeak::Observable;
    ///
    /// let mut name = Observable::new(String::from("DefaultName"));
    /// name.mutate(|n| n.push_str("X"));
    /// name.mutate(|n| n.push_str("Y"));
    /// name.mutate(|n| n.push_str("Z"));
    ///
    /// assert_eq!(name.as_str(), "DefaultNameXYZ");
    /// ```
    pub fn mutate<M>(&mut self, mutation: M)
    where
        M: FnOnce(&mut T),
    {
        mutation(&mut self.value);
        self.delegate.broadcast(&self.value);
    }
}

impl<T> Default for Observable<'_, T>
where
    T: Default,
{
    fn default() -> Self {
        Self::new(Default::default())
    }
}

impl<T> Deref for Observable<'_, T> {
    type Target = T;
    fn deref(&self) -> &Self::Target {
        &self.value
    }
}


src/lib.rs
//! This library provides types allowing execution of callbacks in response to
//! values being broadcast or mutated.
//!
//! <br>
//!
//! # Manually broadcast values
//!
//! Delegates can be used to manually broadcast values to subscribers:
//!
//! ```rust
//! use squeak::{Delegate, Response};
//!
//! let on_damage_received = Delegate::new();
//! on_damage_received.subscribe(|amount| {
//!     println!("Received {amount} damage");
//!     Response::StaySubscribed
//! });
//!
//! on_damage_received.broadcast(16); // Prints "Received 16 damage"
//! on_damage_received.broadcast(14); // Prints "Received 14 damage"
//! on_damage_received.broadcast(28); // Prints "Received 28 damage"
//! ```
//!
//! # Automatically broadcast when a variable is mutated
//!
//! Observables own a value and execute callbacks whenever the value is mutated:
//!
//! ```rust
//!
//! use squeak::{Observable, Response};
//!
//! let mut health = Observable::new(100);
//! health.subscribe(|updated_health| {
//!     println!("Health is now {updated_health}");
//!     Response::StaySubscribed
//! });
//!
//! health.mutate(|h| *h -= 10); // Prints "Health is now 90"
//! health.mutate(|h| *h -= 5);  // Prints "Health is now 85"
//! health.mutate(|h| *h += 25); // Prints "Health is now 110"
//! ```
//!
#![no_std]
extern crate alloc;

mod delegate;
mod observable;

pub use delegate::{Delegate, Response, Subscription};
pub use observable::Observable;
